src/clients/c++/examples/simple_cuda_shm_client.cc

```diff
@@ -0,0 +1,497 @@
+  int output_byte_size =
+      infer_ctx->ByteSize(output0->Dims(), ni::DataType::TYPE_INT32);
+  size cuda_ipc_byte_size = sizeof(ipcCUDA_t);
+
+  // Create Output0 and Output1 in CUDA Shared Memory and store handle in system
+  // shared memory
+  std::string shm_key = "/output_simple";
+  // shared memory for CUDA IPC memory and event handlers
+  int shm_fd_op = CreateSharedMemoryRegion(shm_key, cuda_ipc_byte_size * 2);
+  ipcCUDA_t* output0_cuda_rep =
+      (ipcCUDA_t*)(MapSharedMemory(shm_fd_op, 0, cuda_ipc_byte_size);
+  ipcCUDA_t* output1_cuda_rep =
+      (ipcCUDA_t*)(MapSharedMemory(shm_fd_op, cuda_ipc_byte_size, cuda_ipc_byte_size);
+
+  // Register Output shared memory with TRTIS
+  err = shared_memory_ctx->RegisterSharedMemory(
+      "output_data", "/output_simple", 0, cuda_ipc_byte_size * 2);
+  if (!err.IsOk()) {
+    std::cerr << "error: unable to register shared memory output region: "
+              << err << std::endl;
+    exit(1);
+  }
+
+  // Set the context options to do batch-size 1 requests. Also request that
+  // all output tensors be returned using shared memory.
+  std::unique_ptr<nic::InferContext::Options> options;
+  FAIL_IF_ERR(
+      nic::InferContext::Options::Create(&options),
+      "unable to create inference options");
+
+  options->SetBatchSize(1);
+  options->AddSharedMemoryResult(output0, "output_data", 0, cuda_ipc_byte_size);
+  options->AddSharedMemoryResult(
+      output1, "output_data", cuda_ipc_byte_size, cuda_ipc_byte_size);
+
+  FAIL_IF_ERR(
+      infer_ctx->SetRunOptions(*options), "unable to set inference options");
+
+  // Create Output0 and Output1 in CUDA Shared Memory and store handle in system shared memory. Initialize Input0 to unique
+  // integers and Input1 to all ones.
+  shm_key = "/input_simple";
+  int shm_fd_ip = CreateSharedMemoryRegion(shm_key, cuda_ipc_byte_size * 2);
+  ipcCUDA_t* input0_cuda_rep =
+      (ipcCUDA_t*)(MapSharedMemory(shm_fd_ip, 0, cuda_ipc_byte_size);
+  ipcCUDA_t* input1_cuda_rep =
+      (ipcCUDA_t*)(MapSharedMemory(shm_fd_ip, cuda_ipc_byte_size, cuda_ipc_byte_size);
+
+  // Initialize data to load into shared memory region
+  int input0_data[16];
+  int input1_data[16];
+  for (size_t i = 0; i < 16; ++i) {
+    input0_data[i] = i;
+    input1_data[i] = 1;
+  }
+
+  // copy INPUT0 and INPUT1 data onto GPU
+  int *input0_d_ptr, *input1_d_ptr;
+  checkCudaErrors(cudaMalloc((void**)&d_ptr, 16 * sizeof(int)));
+  checkCudaErrors(cudaMemcpy(
+      (void*)d_ptr, (void*)input_data, 16 * sizeof(int),
+      cudaMemcpyHostToDevice));
+
+  checkCudaErrors(cudaMalloc((void**)&d_ptr, 16 * sizeof(int)));
+  checkCudaErrors(cudaMemcpy(
+      (void*)d_ptr, (void*)input_data, 16 * sizeof(int),
+      cudaMemcpyHostToDevice));
+
+  CreateCUDAIPCHandle(input0_cuda_rep, input0_d_ptr);
+  CreateCUDAIPCHandle(input1_cuda_rep, input1_d_ptr);
+
+  // Register Input shared memory with TRTIS
+  err = shared_memory_ctx->RegisterSharedMemory(
```

Comment:
Need to extend the API here or in option so that the server can know what type of shared memory it is working with as input / output

Reply:
Yes, adding a kind field.

Reply:
Not needed anymore due to different Cuda Register function